### DISCUSSION TITLE
What would this library do?

### DISCUSSION THREAD

**Discussion by cball** [echobind/prisma-factory#4] [Announcements]
`prisma-factory` would be used in test in environments to create valid data for tests. It would lean on the types Prisma gives us. Prisma would be a peerDependency.

Factory generation is an [open discussion](https://github.com/echobind/prisma-factory/discussions/1), but the current thought is we start by exposing a `createFactory` function. Consumers of the lib can use this to create factories for their prisma models.

Once a factory is defined, it can be used like so:

```ts
const user1 = await UserFactory.create({ roles: [Role.ADMIN]})

const user2 = await UserFactory.create({
  email: 'hey@wee.net',
  organization: { create: OrganizationFactory.build() }
});

const org2 = await OrganizationFactory.create({
  name: "Amazing Org",
  users: {
    connect: { id: user1.id },
    create: UserFactory.build({
      roles: [Role.USER]
    })
  }
});
```

This is really powerful for creating data for tests.

**Comment by codingmatty** [comment]
Some features I think would help this once you can do simple create/builds (drawing mostly from my experience with [factory_bot](https://github.com/thoughtbot/factory_bot) in rails):

* Transient attributes - to be able to derive a persisted value from an input value
* Traits - to be able to create groups of attributes/relationships and use a single "trait"
* Sequences - to be able to generate a list of objects with unique ids/values

Example: 
```
const adminUser = await UserFactory.create({ traits: ["admin"], attributes: {...} })

const user = await UserFactory.create({
  attributes: {
    email: 'hey@wee.net',
    organization: { create: OrganizationFactory.build() },
    accounts: { createMany: AccountFactory.buildMany(5) }
   },
   transients: { // to generate fullName "John Smith"
    firstName: "John",
    lastName: "Smith"
  }
})
```

I'd be happy to help any way I can. 

I've been using [this library](https://github.com/ratson/factory-bot#readme) recently, and while it's nifty, it was _not_ made with Prisma in mind, and I've had to hack a lot of things together to make it work.

**Reply by cball** [reply]
Solid info, thanks!

I feel like traits would be the most commonly used out of those patterns.

Probably the best place to start once we get the basics shipped.
